File: Python/15-Greedy/Problems/2-GasStation/GasStationPro.py
"""
The "Gas Station" problem:

Imagine you have a circular route with n gas stations. At each gas station i, you can get gas[i] amount of gas.
You have a car with an unlimited gas tank, and it costs cost[i] amount of gas to travel from station i to its next station (i+1).
You begin the journey with an empty tank at one of the gas stations.

Given two integer arrays gas and cost, return the starting gas station's index if you can travel around the circuit once in the clockwise direction; otherwise, return -1.
If there exists a solution, it is guaranteed to be unique.

Constraints:
    n == gas.length == cost.length
    1 <= n <= 10^4
    0 <= gas[i], cost[i] <= 10^4

Example 1:
    Input: gas = [1,2,3,4,5], cost = [3,4,5,1,2]
    Output: 3
    Explanation:
    Start at the gas station with index 3 (starting fuel = 0).
    - Travel to station 4. Your tank = 0 + 4 - 1 = 3
    - Travel to station 0. Your tank = 3 + 5 - 2 = 6
    - Travel to station 1. Your tank = 6 + 1 - 3 = 4
    - Travel to station 2. Your tank = 4 + 2 - 4 = 2
    - Travel to station 3. Your tank = 2 + 3 - 5 = 0
    You can complete the circuit rotationally, so return 3.

Example 2:
    Input: gas = [2,3,4], cost = [3,4,3]
    Output: -1
    Explanation:
    You can't start at any gas station to complete the circuit.

"""
from typing import List

def can_complete_circuit_brute_force(gas: List[int], cost: List[int]) -> int:
    """
    Approach 1: Brute Force
    - Try every station as a starting point.
    - Simulate the journey and check if it's possible to complete the circuit.
    - This is the most straightforward but least efficient approach.

    Time Complexity: O(n^2), where n is the number of gas stations.
    Space Complexity: O(1)
    """
    n = len(gas)
    for start_station in range(n):
        current_gas = 0
        possible = True
        for i in range(n):
            # Calculate the current station index in the circular route.
            current_station = (start_station + i) % n
            current_gas += gas[current_station] - cost[current_station]
            if current_gas < 0:
                possible = False
                break
        if possible:
            print(f"Brute Force: Starting station = {start_station}")
            return start_station
    print("Brute Force: No solution")
    return -1


def can_complete_circuit_optimized(gas: List[int], cost: List[int]) -> int:
    """
    Approach 2: Optimized Single Pass
    - If the total gas is less than the total cost, there is no solution.
    - Start from station 0, and keep track of the current gas and the starting station.
    - If the current gas becomes negative, change the starting station to the next station.
    - This approach is based on the observation that if we cannot reach station j from station i,
      then we cannot reach station j from any station between i and j.

    Time Complexity: O(n), where n is the number of gas stations.
    Space Complexity: O(1)
    """
    n = len(gas)
    total_gas = sum(gas)
    total_cost = sum(cost)

    if total_gas < total_cost:
        print("Optimized Single Pass: No solution")
        return -1

    start_station = 0
    current_gas = 0
    for i in range(n):
        current_gas += gas[i] - cost[i]
        if current_gas < 0:
            start_station = i + 1
            current_gas = 0

    print(f"Optimized Single Pass: Starting station = {start_station}")
    return start_station

def can_complete_circuit_kadane(gas: List[int], cost: List[int]) -> int:
    """
    Approach 3: Kadane's Algorithm Variation
    - Similar to the optimized single pass, but uses a variation of Kadane's algorithm.
    - Kadane's algorithm is typically used to find the maximum subarray sum.
    - Here, we adapt it to find the starting point with the maximum "gas surplus".

    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    n = len(gas)
    total_gas = 0
    current_gas = 0
    start_station = 0

    for i in range(n):
        total_gas += gas[i] - cost[i]
        current_gas += gas[i] - cost[i]
        if current_gas < 0:
            current_gas = 0
            start_station = i + 1

    if total_gas < 0:
        print("Kadane Variation: No solution")
        return -1
    print(f"Kadane Variation: Starting station = {start_station}")
    return start_station
def can_complete_circuit_deque(gas: List[int], cost: List[int]) -> int:
    """
    Approach 4: Using Deque (Double-Ended Queue)
    - Extend the gas and cost arrays to simulate the circular route.
    - Use a deque to maintain a window of possible starting points.
    - If a starting point is not feasible, remove it from the deque.

    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    from collections import deque

    n = len(gas)
    extended_gas = gas + gas
    extended_cost = cost + cost
    q = deque(range(n))  # Initialize deque with possible starting stations

    for i in range(2 * n):
        while q and sum(extended_gas[q[0]:i+1]) < sum(extended_cost[q[0]:i+1]):
            q.popleft()  # Remove infeasible starting stations
        if not q:
            print("Deque Approach: No solution")
            return -1
        if i >= n and q[0] < n:
            print(f"Deque Approach: Starting station = {q[0]}")
            return q[0]
    print("Deque Approach: No solution")
    return -1

def can_complete_circuit_dynamic_programming(gas: List[int], cost: List[int]) -> int:
    """
    Approach 5: Dynamic Programming (Less Efficient, but Demonstrates DP)
    - Create a DP table where dp[i] stores the remaining gas at station i, starting from station 0.
    - Iterate through all possible starting stations and use the DP table to check for completion.
    - While DP can be used, it's not the most efficient approach for this problem.

    Time Complexity: O(n^2)
    Space Complexity: O(n)
    """
    n = len(gas)
    for start_station in range(n):
        dp = [0] * n
        dp[0] = gas[start_station] - cost[start_station]
        possible = True
        for i in range(1, n):
            current_station = (start_station + i) % n
            prev_station = (start_station + i - 1) % n
            dp[i] = dp[i - 1] + gas[current_station] - cost[current_station]
            if dp[i] < 0:
                possible = False
                break
        if possible and dp[-1] >= 0: # Check if the tank has >=0 gas at the end
            print(f"Dynamic Programming: Starting station = {start_station}")
            return start_station
    print("Dynamic Programming: No solution")
    return -1

def can_complete_circuit_greedy(gas: List[int], cost: List[int]) -> int:
    """
    Approach 6: Greedy Approach (Most Efficient)
    - This is the most efficient and intuitive solution.
    - The core idea is that if the total gas is greater than or equal to the total cost,
      there must be a solution.
    - We maintain a running `tank` and a `start` index.  If the `tank` goes negative,
      we reset the `start` to the next station.

    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    n = len(gas)
    tank = 0
    total_gas = 0
    total_cost = 0
    start = 0

    for i in range(n):
        tank += gas[i] - cost[i]
        total_gas += gas[i]
        total_cost += cost[i]
        if tank < 0:
            start = i + 1
            tank = 0

    if total_gas < total_cost:
        print("Greedy Approach: No solution")
        return -1  # Not possible to complete the circuit
    else:
        print(f"Greedy Approach: Starting station = {start}")
        return start

if __name__ == "__main__":
    gas = [1, 2, 3, 4, 5]
    cost = [3, 4, 5, 1, 2]
    #gas = [2, 3, 4]
    #cost = [3, 4, 3]

    print("Gas and Cost:", gas, cost)
    print("-" * 30)
    print("Brute Force Result:", can_complete_circuit_brute_force(gas, cost))
    print("-" * 30)
    print("Optimized Single Pass Result:", can_complete_circuit_optimized(gas, cost))
    print("-" * 30)
    print("Kadane Variation Result:", can_complete_circuit_kadane(gas, cost))
    print("-" * 30)
    print("Deque Approach Result:", can_complete_circuit_deque(gas, cost))
    print("-" * 30)
    print("Dynamic Programming Result:", can_complete_circuit_dynamic_programming(gas, cost))
    print("-" * 30)
    print("Greedy Approach Result:", can_complete_circuit_greedy(gas, cost))
    print("-" * 30)

File: C++/15-Greedy/Problems/2-GasStation/GasStationPro.js
/*
The "Gas Station" problem:

You are given an array of gas stations along a circular route, where gas[i] is the amount of gas available at the i-th station, and cost[i] is the amount of gas required to travel from the i-th station to the (i+1)-th station.  Determine the starting gas station index if you can complete one full circuit.

Problem Variations and Real-World Applications:

Here are five variations of the Gas Station problem, along with real-world applications and JavaScript implementations:

1.  **Basic Gas Station Problem:**
    * **Description:** Find a starting station to complete the circuit.
    * **Real-World Application:** Route planning for delivery trucks with fuel constraints.
*/
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
const canCompleteCircuit = function (gas, cost) {
    const n = gas.length;
    let totalGas = 0;
    let currentGas = 0;
    let startingStation = 0;

    for (let i = 0; i < n; i++) {
        totalGas += gas[i] - cost[i];
        currentGas += gas[i] - cost[i];
        if (currentGas < 0) {
            startingStation = i + 1;
            currentGas = 0;
        }
    }

    return totalGas >= 0 ? startingStation : -1;
};

// Example usage:
const gas1 = [1, 2, 3, 4, 5];
const cost1 = [3, 4, 5, 1, 2];
console.log("Basic Gas Station Problem:");
console.log(canCompleteCircuit(gas1, cost1)); // Output: 3

const gas2 = [2, 3, 4];
const cost2 = [3, 4, 3];
console.log(canCompleteCircuit(gas2, cost2)); // Output: -1

/*
2.  **Gas Station with Limited Tank Capacity:**
    * **Description:** You have a limited fuel tank capacity.  Can you complete the circuit?
    * **Real-World Application:** Planning routes for drones with limited battery capacity.
*/
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @param {number} tankCapacity
 * @return {boolean}
 */
const canCompleteCircuitWithTank = function (gas, cost, tankCapacity) {
    const n = gas.length;
    let currentGas = 0;
    let startingStation = 0;
    let traveledStations = 0; // Keep track of how many stations we've visited

    for (let i = 0; i < n * 2; i++) { // Iterate twice the length to simulate circularity
        const stationIndex = i % n; // Use modulo to handle circular array
        currentGas += gas[stationIndex];

        if (currentGas < cost[stationIndex]) {
            // Cannot reach the next station
            if (traveledStations >= n) return false; // Circuit cannot be completed
            startingStation = stationIndex + 1;
            currentGas = 0;
            traveledStations = 0; // Reset count when starting from new station
        } else {
            currentGas -= cost[stationIndex];
            traveledStations++;
            if (currentGas > tankCapacity) {
                return false; // Exceeds tank capacity
            }
            if (traveledStations >= n) {
                return true; // Completed the circuit
            }
        }
    }
    return false; //cannot complete
};

// Example usage:
const gas3 = [1, 2, 3, 4, 5];
const cost3 = [3, 4, 5, 1, 2];
const tankCapacity1 = 5;
console.log("\nGas Station Problem with Limited Tank:");
console.log(canCompleteCircuitWithTank(gas3, cost3, tankCapacity1)); // Output: true (Starting at 3, gas at each station: 4, 6, 9, 8, 6. Cost: 1, 2, 3, 4, 5)

const gas4 = [2, 3, 4];
const cost4 = [3, 4, 3];
const tankCapacity2 = 4;
console.log(canCompleteCircuitWithTank(gas4, cost4, tankCapacity2)); // Output: false

const gas5 = [1, 5, 3, 4, 5];
const cost5 = [3, 4, 5, 1, 2];
const tankCapacity3 = 6;
console.log(canCompleteCircuitWithTank(gas5, cost5, tankCapacity3)); //Output: false

/*
3.  **Gas Station with Variable Costs:**
    * **Description:** The cost of gas varies at each station. Find the cheapest starting point.
    * **Real-World Application:** Optimizing refueling stops for long-haul trucks to minimize fuel expenses.
*/

/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @param {number[]} pricePerGallon
 * @return {number}
 */
const findCheapestStartingStation = function (gas, cost, pricePerGallon) {
    const n = gas.length;
    let minCost = Infinity;
    let cheapestStartingStation = -1;

    for (let start = 0; start < n; start++) {
        let currentGas = 0;
        let currentCost = 0;
        let possible = true;

        for (let i = 0; i < n; i++) {
            const stationIndex = (start + i) % n;
            currentGas += gas[stationIndex];
            if (currentGas < cost[stationIndex]) {
                possible = false;
                break;
            }
            currentGas -= cost[stationIndex];
            currentCost += cost[stationIndex] * pricePerGallon[stationIndex];
        }

        if (possible && currentCost < minCost) {
            minCost = currentCost;
            cheapestStartingStation = start;
        }
    }

    return cheapestStartingStation;
};

// Example usage:
const gas6 = [1, 2, 3, 4, 5];
const cost6 = [3, 4, 5, 1, 2];
const pricePerGallon1 = [2, 3, 4, 1, 2]; // Price of gas at each station
console.log("\nGas Station Problem with Variable Costs:");
console.log(findCheapestStartingStation(gas6, cost6, pricePerGallon1)); // Output: 3

const gas7 = [2, 3, 4];
const cost7 = [3, 4, 3];
const pricePerGallon2 = [5, 2, 1];
console.log(findCheapestStartingStation(gas7, cost7, pricePerGallon2)); // Output: -1

const gas8 = [1, 2, 3, 4, 5];
const cost8 = [3, 4, 5, 1, 2];
const pricePerGallon3 = [2, 3, 4, 1, 5];
console.log(findCheapestStartingStation(gas8, cost8, pricePerGallon3));
/*
4.  **Gas Station with Two-Way Roads:**
    * **Description:** You can travel between stations in both directions. Find the shortest path.
    * **Real-World Application:** Optimizing routes in a city with two-way streets and varying gas prices.
*/
/**
 * @param {number[]} gas
 * @param {number[][]} costMatrix // costMatrix[i][j] is cost from i to j
 * @return {number[]} [startingStation, minCost]
 */
const findShortestPathTwoWay = function (gas, costMatrix) {
    const n = gas.length;
    let minCost = Infinity;
    let shortestPath = [-1, -1];

    for (let start = 0; start < n; start++) {
        let currentGas = 0;
        let currentCost = 0;
        let possible = true;
        const visited = new Array(n).fill(false);
        let currentStation = start;
        let stationsVisited = 0;

        while (stationsVisited < n) {
            visited[currentStation] = true;
            stationsVisited++;
            let nextStation = -1;
            let minNextCost = Infinity;

            for (let j = 0; j < n; j++) {
                if (!visited[j]) {
                    if (currentGas + gas[currentStation] >= costMatrix[currentStation][j]) {
                        if (costMatrix[currentStation][j] < minNextCost) {
                            minNextCost = costMatrix[currentStation][j];
                            nextStation = j;
                        }
                    }
                }
            }
            if (nextStation === -1) {
                possible = false;
                break;
            }
            currentGas += gas[currentStation] - costMatrix[currentStation][nextStation];
            currentCost += costMatrix[currentStation][nextStation];
            currentStation = nextStation;
        }
        if (possible && currentStation === start && currentCost < minCost) {
            minCost = currentCost;
            shortestPath = [start, minCost];
        }
    }
    return shortestPath;
};

// Example usage:
const gas9 = [10, 2, 3, 6, 5];
const costMatrix1 = [
    [0, 2, 10, 12, 5],
    [2, 0, 5, 8, 4],
    [10, 5, 0, 3, 9],
    [12, 8, 3, 0, 6],
    [5, 4, 9, 6, 0]
];
console.log("\nGas Station Problem with Two-Way Roads:");
console.log(findShortestPathTwoWay(gas9, costMatrix1));

const gas10 = [1, 2, 3];
const costMatrix2 = [
    [0, 5, 10],
    [5, 0, 3],
    [10, 3, 0]
];
console.log(findShortestPathTwoWay(gas10, costMatrix2));

/*
5. **Gas Station with Service Time:**
* **Description**: Each gas station has a service time.  Find the optimal starting station to minimize total time.
* **Real-world application**: Route optimization for emergency vehicles (ambulances, fire trucks) where time is critical.
*/

/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @param {number[]} serviceTime
 * @return {number}
 */
const findOptimalStartingStationWithServiceTime = function (gas, cost, serviceTime) {
    const n = gas.length;
    let minTotalTime = Infinity;
    let optimalStartingStation = -1;

    for (let start = 0; start < n; start++) {
        let currentGas = 0;
        let totalTime = 0;
        let possible = true;

        for (let i = 0; i < n; i++) {
            const stationIndex = (start + i) % n;
            currentGas += gas[stationIndex];
            if (currentGas < cost[stationIndex]) {
                possible = false;
                break;
            }
            currentGas -= cost[stationIndex];
            totalTime += cost[stationIndex] + serviceTime[stationIndex];
        }

        if (possible && totalTime < minTotalTime) {
            minTotalTime = totalTime;
            optimalStartingStation = start;
        }
    }

    return optimalStartingStation;
};

// Example Usage
const gas11 = [1, 2, 3, 4, 5];
const cost11 = [3, 4, 5, 1, 2];
const serviceTime1 = [2, 3, 1, 4, 2];
console.log("\nGas Station Problem with Service Time:");
console.log(findOptimalStartingStationWithServiceTime(gas11, cost11, serviceTime1));

const gas12 = [2, 3, 4];
const cost12 = [3, 4, 3];
const serviceTime2 = [1, 2, 1];
console.log(findOptimalStartingStationWithServiceTime(gas12, cost12, serviceTime2));

const gas13 = [10, 2, 5, 4, 5];
const cost13 = [3, 4, 5, 1, 2];
const serviceTime3 = [5, 1, 2, 3, 1];
console.log(findOptimalStartingStationWithServiceTime(gas13, cost13, serviceTime3));
